File: src/db/bedrock-db-data/load_assets/load_assets.py
import os
import json
import csv
import sys
import psycopg2

def usage():
    print('  Usage:  load_assets database-host [data-directory]')

if len(sys.argv) < 2:
    usage()
    sys.exit(2)

db_host = sys.argv[1].split(':')[0]
db_name = 'bedrock'
db_user = 'bedrock'
db_password = 'test-bedrock'

data_directory = '../data'
if (len(sys.argv) > 2):
  data_directory = sys.argv[2]

conn = psycopg2.connect(
    host = db_host,
    dbname = db_name,
    password = db_password,
    user = db_user
  )
cur = conn.cursor()

print("Database host is ", db_host)
print("Data directory is ", data_directory)

# clear out old data
sql = f'''
  truncate table bedrock.assets;
  truncate table bedrock.dependencies;
  truncate table bedrock.etl;
  truncate table bedrock.tasks;
  truncate table bedrock.asset_tags;
  truncate table bedrock.tags;
  truncate table bedrock.run_groups;
  truncate table bedrock.connections;
  truncate table bedrock.custom_fields;
  truncate table bedrock.custom_values;
'''
cur.execute(sql)
print('Truncated all tables')

# Load the run groups
sql = 'INSERT INTO bedrock.run_groups (run_group_name,	cron_string) VALUES '
runGroups = []
with open(os.path.join(data_directory,'run_groups.csv')) as ff:
  rdr = csv.reader(ff)
  
  i = 0;
  rows = list(rdr)
  nrows = len(rows)
  for row in rows:
    i = i+1
    sql = f"{sql} ('{row[0]}', '{row[1]}')"
    if (i<nrows):
      sql = sql + ','
cur.execute(sql)
print(f'Wrote {nrows} items to the run_groups table')

# Load the tags
sql = 'INSERT INTO bedrock.tags (tag_name,	display_name) VALUES '
tags = []
with open(os.path.join(data_directory,'tags.csv')) as ff:
  rdr = csv.reader(ff)
  
  i = 0;
  rows = list(rdr)
  nrows = len(rows)
  for row in rows:
    i = i+1
    sql = f"{sql} ('{row[0]}', '{row[1]}')"
    if (i<nrows):
      sql = sql + ','
cur.execute(sql)
print(f'Wrote {nrows} items to the tags table')

# Load the connections
sql = 'INSERT INTO bedrock.connections (connection_name, description, connection_class) VALUES '
tags = []
with open(os.path.join(data_directory,'connections.csv')) as ff:
  rdr = csv.reader(ff)
  
  i = 0;
  rows = list(rdr)
  nrows = len(rows)
  for row in rows:
    i = i+1
    sql = f"{sql} ('{row[0]}', '{row[1]}', '{row[2]}')"
    if (i<nrows):
      sql = sql + ','
cur.execute(sql)
print(f'Wrote {nrows} items to the connections table')

# Load the custom fields
customs_map = {}
sql = 'INSERT INTO bedrock.custom_fields (asset_type, field_name,	field_display, field_type) VALUES '
with open(os.path.join(data_directory,'custom_fields.csv')) as ff:
  rdr = csv.reader(ff)
  
  i = 0;
  rows = list(rdr)
  nrows = len(rows)
  for row in rows:
    i = i+1
    sql = f"{sql} ('{row[0]}', '{row[1]}', '{row[2]}', '{row[3]}')"
    if row[0] not in customs_map:
      customs_map[row[0]] = []
    customs_map[row[0]].append(row[1]);
    if (i<nrows):
      sql = sql + ','
cur.execute(sql)
print(f'Wrote {nrows} items to the custom fields table')

# Load the assets
print('Load all assets')
assets_directory = data_directory + '/assets'
for asset_subdir in os.listdir(assets_directory):
  # each asset
  print('  Processing ', asset_subdir)
  d = os.path.join(assets_directory, asset_subdir)
  if os.path.isdir(d):
    for file in os.listdir(d):
      if file.endswith('.ETL.json') or file.endswith('.etl.json'):
        # etlFile
        with open(os.path.join(d, file), 'r') as ff:
          etl = json.load(ff)
          asset_name = etl['asset_name']
          sql = f'''
          insert into bedrock.etl(asset_name, run_group, active)
          values(%s, %s, %s);
          '''
          cur.execute(sql,(asset_name, etl['run_group'], "true"))
          tasks = etl['tasks']
          for seq_number, task in enumerate(tasks):
            type = task['type']
            if 'description' in task:
              description = task['description']
            else:
              description = None
            if type == "table_copy" or type == "file_copy" or type == "aggregate":
              # table_copy / file_copy
              sql = f'''
              insert into bedrock.tasks(asset_name, seq_number, description, type, active, source, target, configuration)
              values(%s, %s, %s, %s, %s, %s, %s, %s);
              '''
              cur.execute(sql, (asset_name, seq_number, description, type, task['active'], 
              json.dumps(task['source_location']), json.dumps(task['target_location']), None))
            elif type == "sql":
              # sql
              sqlstring = task['sql_string']
              sql = f'''
              insert into bedrock.tasks(asset_name, seq_number, description, type, active, source, target, configuration)
              values(%s, %s, %s, %s, %s, %s, %s, %s);
              '''
              cur.execute(sql, (asset_name, seq_number, description, type, task['active'], 
              None, json.dumps({ "connection": task['connection'] }), sqlstring))
            else:
              # everything else
              sql = f'''
              insert into bedrock.tasks(asset_name, seq_number, description, type, active, source, target, configuration)
              values(%s, %s, %s, %s, %s, %s, %s, %s);
              '''
              cur.execute(sql, (asset_name, seq_number, description, type, task['active'], 
              None, json.dumps(task), None))
      elif file.endswith('.json'):
        # configFile
        with open(os.path.join(d, file), 'r') as ff:
          config = json.load(ff)
          asset_name = config['asset_name']
          sql = f'''
            insert into bedrock.assets
            (asset_name, description, location, owner_id, notes, active)
            values(%s, %s, %s, %s, %s, %s);
          '''
          cur.execute(sql,
          (asset_name, config["description"], json.dumps(config["location"]), config["owner_id"], config["notes"], config["active"]
          ))
          
          # custom variables
          if config['asset_type'] is not None and 'asset_type' in config and config['asset_type'] in customs_map:

            for itm in customs_map[config['asset_type']]:
              if itm in config:
                sql = f'''
                  insert into bedrock.custom_values
                  (asset_name, field_name, field_value)
                  values(%s, %s, %s);
                '''
                cur.execute(sql, (asset_name, itm, config[itm]))

          # dependencies
          dependencies = config['depends']         
          if dependencies:
            for depend in dependencies:
              sql = f'''
              insert into bedrock.dependencies(asset_name, dependency)
              values(%s, %s);
              '''
              cur.execute(sql, (asset_name, depend))
          
          # asset tags
          tags = config['tags']
          if tags:
            for tag in tags:
              sql = f'''
              insert into bedrock.asset_tags(asset_name, tag_name)
              values(%s, %s);
              '''
              cur.execute(sql, (asset_name, tag))              

    # nm = cur.fetchone()[0] # probably can skip this, but maybe to double check

conn.commit()
cur.close()
conn.close()

File: src/db/bedrock-db-data/load_assets/load_assets.js
// A ONE TIME LOAD: Create JSON files from DB
/* eslint-disable no-await-in-loop */
/* eslint-disable no-console */
import pgpkg from 'pg';
const { Client } = pgpkg;
import { existsSync, mkdirSync, readFileSync, readdirSync } from 'fs';

import { getDBConnection } from 'bedrock_common';
import { parse } from 'csv-parse/sync';

async function clearOutDB(client) {
  let sql = `
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.assets;
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.dependencies;
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.etl;
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.tasks;
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.asset_tags;
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.tags;
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.run_groups;
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.connections;
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.custom_fields;
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.custom_values;
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.owners;
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.asset_types;
    truncate table ${process.env.BEDROCK_DB_SCHEMA}.asset_type_custom_fields;
  `;
  await client.query(sql)
    .catch((err) => {
      const errmsg = [err.code];
      throw new Error([`Postgres error: ${errmsg}`, err]);
    });
}


async function readCsvFiles(client, data_directory) {
  let file_names = readdirSync(data_directory);
  for (let file_name of file_names) {
    let fileNm = file_name.split('.');
    let tableName = fileNm[0];
    if (fileNm[fileNm.length - 1].toLowerCase() === 'csv')  {
      let fileContent = readFileSync  (data_directory + '/' + file_name, 'utf8').trim();

      let values = parse(fileContent,{ 
        cast: (value, {quoting}) => {
            if (value === 'NULL') {return null;}
            if (quoting) {return value;}
            if (value === '') {return null;}
            return value;
          }
       });
      let numCols = values[0].length;

      // load into DB
      let sql = `insert into ${process.env.BEDROCK_DB_SCHEMA}.${tableName} values (`;
      let ii = []; // for the $1, $2, $3, etc.
      for (let i = 1; i < numCols + 1; i++) {
        ii.push(`$${i}`);
      }
      sql += ii.join(',') + ');';
      for(let row of values) {
        await client.query(sql, row)
          .catch((err) => {
            const errmsg = [err.code];
            throw new Error([`Postgres error: ${errmsg}`, err]);
          });
      }
    }
  }
}

async function readAssetFiles(client, assets_directory) {
  if (existsSync(assets_directory)) {
    let asset_subdirs = readdirSync(assets_directory);
    for (let asset_subdir of asset_subdirs) {
      let file_list = readdirSync(assets_directory + asset_subdir);
      for (let f of file_list) {
        let fileContent = readFileSync(assets_directory + asset_subdir + '/' + f, 'utf8');
        let fileNm = f.split('.');
        if (fileNm[fileNm.length - 2].toLowerCase() === 'etl')  {
          // ETL files
            let etl = JSON.parse(fileContent);
            let sql = `insert into ${process.env.BEDROCK_DB_SCHEMA}.etl(asset_id, run_group_id, active) values ($1,$2,$3);`;
            await client.query(sql, [etl.asset_id, etl.run_group_id, etl.active])
              .catch((err) => {
                const errmsg = [err.code];
                throw new Error([`Postgres error: ${errmsg}`, err]);
              });
            for (let task of etl.tasks) {
              sql = `insert into ${process.env.BEDROCK_DB_SCHEMA}.tasks
              (task_id, asset_id, seq_number, description, "type", active, "source", target, "configuration") values ($1,$2,$3,$4,$5,$6,$7,$8,$9);`;
              await client.query(sql, [task.task_id, etl.asset_id, task.seq_number, task.description, task.type, task.active, task.source, 
                task.target, task.configuration])
                .catch((err) => {
                  const errmsg = [err.code];
                  throw new Error([`Postgres error: ${errmsg}`, err]);
                });
            }
        } else {
          // Asset files
          let asset = JSON.parse(fileContent);
          // console.log(asset);
          let sql = `insert into ${process.env.BEDROCK_DB_SCHEMA}.assets
          (asset_id, asset_name, description, "location", asset_type_id, owner_id, notes, link, active) values ($1,$2,$3,$4,$5,$6,$7,$8,$9);`;
          await client.query(sql, [asset.asset_id, asset.asset_name, asset.description, asset.location, asset.asset_type_id, asset.owner_id, asset.notes, asset.link, asset.active])
            .catch((err) => {
              const errmsg = [err.code];
              throw new Error([`Postgres error: ${errmsg}`, err]);
            });
          for (let tag of asset.tags) {
            sql = `insert into ${process.env.BEDROCK_DB_SCHEMA}.asset_tags(asset_id, tag_id) values ($1,$2);`;
            await client.query(sql, [asset.asset_id, tag])
              .catch((err) => {
                const errmsg = [err.code];
                throw new Error([`Postgres error: ${errmsg}`, err]);
              });
          }
          for (let dependency of asset.depends) {
            sql = `insert into ${process.env.BEDROCK_DB_SCHEMA}.dependencies(asset_id, dependent_asset_id) values ($1,$2);`;
            await client.query(sql, [asset.asset_id, dependency])
              .catch((err) => {
                const errmsg = [err.code];
                throw new Error([`Postgres error: ${errmsg}`, err]);
              });
          }
          if(asset.custom_fields) {
            for (let custom_field of Object.keys(asset.custom_fields)) {
              sql = `insert into ${process.env.BEDROCK_DB_SCHEMA}.custom_values(asset_id, custom_field_id, field_value) values ($1,$2,$3);`;
              await client.query(sql, [asset.asset_id, custom_field, asset.custom_fields[custom_field]])
                .catch((err) => {
                  const errmsg = [err.code];
                  throw new Error([`Postgres error: ${errmsg}`, err]);
                });
            }
          }
        }
      };
    };
  }
}

////////////////////////////////////////////

let data_directory = '../data';
let assets_directory = data_directory + '/assets/';

console.log('Connect to the DB');
const dbConnection = await getDBConnection();
const client = new Client(dbConnection);
await client.connect();
console.log('Connected, read the files');

await clearOutDB(client);
await readCsvFiles(client, data_directory);
await readAssetFiles(client, assets_directory);

await client.end();
